### PR TITLE
Specify shared info and aggregation service payloads for aggregatable report

### PR DESCRIPTION
**Pull request by linnan-github**
***
<a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/pull/481.html" title="Last updated on Nov 30, 2022, 8:38 PM UTC (22cdbcb)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/481/9da4aba...22cdbcb.html" title="Last updated on Nov 30, 2022, 8:38 PM UTC (22cdbcb)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+
+Note: The user agent might enforce weekly key rotation. If there are multiple keys, the user agent
+might independently pick a key uniformly at random for every encryption operation.
+
+An [=aggregatable report=] |report|'s <dfn for="aggregatable report">plaintext payload</dfn>
+is the result of running the following steps:
+
+1. Let |payloadData| be a new [=list/is empty|empty=] [=list=].
+1. [=list/iterate|For each=] |contribution| of |report|'s [=aggregatable report/contributions=]:
+    1. Let |contributionData| be a [=map=] of the following key/value pairs:
```
Comment (alexmturner) [inline index.bs:1640]: Should this be ordered map, to match earlier in the spec?

File: index.bs [hunk 2]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+1. If |report|'s [=aggregatable report/debug mode=] is <strong>enabled</strong>,
+    set |sharedInfo|["`debug_mode`"] to "`enabled`".
+1. Return the [=string=] resulting from executing [=serialize an infra value to a json string=] on |sharedInfo|.
+
+<h3 id="obtain-aggregatable-report-aggregation-service-payloads">Obtaining an aggregatable report's aggregation service payloads</h3>
+
+To <dfn>obtain the public key for encryption</dfn>, return a user agent determined public key asynchronously.
+
+Note: The user agent might enforce weekly key rotation. If there are multiple keys, the user agent
+might independently pick a key uniformly at random for every encryption operation.
```
Comment (alexmturner) [inline index.bs:1633]: optional: might want to briefly discuss the key ID concept?

**Review by alexmturner — COMMENTED** [review]
Sorry for the delay! Got halfway through and then forgot to finish. Generally looks good -- mostly small comments; adding an ability to handle errors might be a bit more complex

File: index.bs [hunk 3]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+
+An [=aggregatable report=] |report|'s <dfn for="aggregatable report">plaintext payload</dfn>
+is the result of running the following steps:
+
+1. Let |payloadData| be a new [=list/is empty|empty=] [=list=].
+1. [=list/iterate|For each=] |contribution| of |report|'s [=aggregatable report/contributions=]:
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:1639]: The order is based on aggregation keys on the source, which is an ordered map.

File: index.bs [hunk 4]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+    :: |payloadData|
+
+1. Return the [=byte sequence=] resulting from [[!RFC8949|CBOR encoding]] |payload|.
+
+To <dfn>obtain the encrypted payload</dfn> given an [=aggregatable report=] |report| and
+a public key |pkR|, run the following steps:
+
+1. Let |plaintext| be |report|'s [=aggregatable report/plaintext payload=].
+1. Let |info| be "`aggregation_service`".
+1. Append |report|'s [=aggregatable report/shared info=] to |info|, [=utf-8 encode|encoded=].
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:1662]: Yeah that looks better, thanks.

File: index.bs [hunk 5]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+    :: |report|'s [=aggregatable report/attribution destination=], <a href="https://html.spec.whatwg.org/multipage/origin.html#serialization-of-a-site">serialized</a>
+    : "`report_id`"
+    :: |report|'s [=aggregatable report/report ID=]
+    : "`reporting_origin`"
+    :: |reportingOrigin|, [=serialization of an origin|serialized=]
+    : "`scheduled_report_time`"
+    :: |report|'s [=aggregatable report/original report time=] in seconds since the UNIX epoch, [=serialize an integer|serialized=]
+    : "`source_registration_time`"
+    :: |report|'s [=aggregatable report/source time=] in seconds since the UNIX epoch, rounded down
+        to a whole day, [=serialize an integer|serialized=]
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:1620]: We have "Round |expiry| away from zero to the nearest day (86400 seconds)." in the spec , so probably its ok?

File: index.bs [hunk 6]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
 Issue: This would ideally be replaced by a more descriptive algorithm in Infra. See
 <a href="https://github.com/whatwg/infra/issues/201">infra/201</a>
 
+<h3 id="encode-integer">Encode an integer</h3>
+
+To <dfn>encode an integer</dfn>, represent it as a big-endian [=byte sequence=], zero-padding.
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:1588]: yeah, that SGTM, thanks.

File: index.bs [hunk 7]
```diff
@@ -1583,6 +1583,107 @@ To <dfn>serialize an integer</dfn>, represent it as a string of the shortest pos
...
+
+To <dfn>obtain the encrypted payload</dfn> given an [=aggregatable report=] |report| and
+a public key |pkR|, run the following steps:
+
+1. Let |plaintext| be |report|'s [=aggregatable report/plaintext payload=].
+1. Let |info| be "`aggregation_service`".
+1. Append |report|'s [=aggregatable report/shared info=] to |info|, [=utf-8 encode|encoded=].
+1. Set up [[RFC9180|HPKE]] [[RFC9180#name-encryption-to-a-public-key|sender's context]]
+    with |pkR| and |info|.
+1. Return the [=byte sequence=] resulting from [[RFC9180#name-encryption-and-decryption|encrypting]]
```
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs:1756]: We don't seem to handle failures in other cases as well, so probably is fine?

[the remark this answers is not in the document]
Comment (alexmturner) [reply index.bs:1756]: Hmm, I'm not sure. My instinct is that we want to specify what happens if the public key fails to be obtained or if it is malformed in some way.

File: index.bs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexmturner) [reply index.bs]: Wdyt about "Rounded down to a multiple of a whole day (86400 seconds)" (or just "whole day multiple")?

File: index.bs (hunk 7 above)
[the remark this answers is not in the document]
Comment (linnan-github) [reply index.bs]: Thanks Alex. Updated with error handling for public key and encryption.

File: index.bs [hunk 8]
```diff
@@ -1726,6 +1829,7 @@ To <dfn>attempt to deliver a report</dfn> given an [=attribution report=] |repor
 
 1. Let |url| be the result of executing [=generate an attribution report URL=] on |report|.
 1. Let |data| be the result of executing [=serialize an attribution report=] on |report|.
+1. If |data| is null, return.
```
Comment (alexmturner) [inline index.bs:1832]: I'm wondering if the later sentence "A user agent may retry this algorithm in the event that there was an error" might need a slight change or to be copied into some of the other algorithms?

Comment (linnan-github) [reply index.bs:1832]: I think this may be fine as the error has already been handled and data being null is not an error?

Comment (alexmturner) [reply index.bs:1832]: Ah I think you're right. This algorithm is the main one that calls all the others. I do wonder if being slightly more explicit about what this looks like could be useful, but not super important -- e.g. "If data is null, wait an implementation-defined amount of time and rerun this algorithm." + something similar in processReponse. (Up to you)

One last thing I spotted -- should we be resetting report's delievered bit on startup?

Comment (linnan-github) [reply index.bs:1832]: Actually updated to return the error to the main algorithm if public key fetching or serialization failed. In this case, the retry statement below that should apply here as well.

Currently the retry logic is not fully spec'ed, so I think it's fine not resetting report's `delivered` bit as we don't want to retry repeatedly.

Comment (alexmturner) [reply index.bs]: Yeah, I think it's ok not to spec the retry behavior. However, it's a little odd to me that the return path neither changes the delivered bit nor "Queues a task to remove report from the aggregatable report cache." Not sure what the best resolution is, though.

File: index.bs [hunk 9]
```diff
@@ -1566,6 +1564,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+    :: A [=string=], API version.
+
+1. If |report|'s [=aggregatable report/debug mode=] is <strong>enabled</strong>,
+    set |sharedInfo|["`debug_mode`"] to "`enabled`".
+1. Return the [=string=] resulting from executing [=serialize an infra value to a json string=] on |sharedInfo|.
+
+<h3 id="obtain-aggregatable-report-aggregation-service-payloads">Obtaining an aggregatable report's aggregation service payloads</h3>
+
+To <dfn>obtain the public key for encryption</dfn>, return a user agent-determined public key asynchronously,
+or an error in the event that the user agent failed to obgtain the public key.
```
[the remark this answers is not in the document]
Comment (alexmturner) [reply index.bs:1916]: (think this was missed)

File: index.bs [hunk 10]
```diff
@@ -2060,6 +2164,7 @@ To <dfn>attempt to deliver a debug report</dfn> given an [=attribution report=]
 1. Let |url| be the result of executing [=generate an attribution report URL=] on |report| with
     [=generate an attribution report URL/isDebugReport=] set to true.
 1. Let |data| be the result of executing [=serialize an attribution report=] on |report|.
+1. If |data| is null, return.
```
Comment (alexmturner) [inline index.bs:2167]: Should this be changed to align? (i.e. s/null/an error/)

File: index.bs (hunk 8 above)
Comment (linnan-github) [reply index.bs]: yeah, that makes sense. Updated to remove the report from the cache before return.

File: index.bs [hunk 11]
```diff
@@ -2132,26 +2132,29 @@ and a [=header list=] |newHeaders| (defaults to an empty [=list=]):
...
+    <dt>[=aggregatable report=]</dt>
+    <dd>[=Queue a task=] to [=list/remove=] |report| from the [=aggregatable report cache=].</dd>
+    </dl>
+
 To <dfn>attempt to deliver a report</dfn> given an [=attribution report=] |report|, run the following steps:
 
 1. Let |url| be the result of executing [=generate an attribution report URL=] on |report|.
 1. Let |data| be the result of executing [=serialize an attribution report=] on |report|.
-1. If |data| is an error, return.
+1. If |data| is an error, run [=remove a report from the cache=] and return.
```
Comment (alexmturner) [inline index.bs:2150]: nit: should this be "run [=remove a report from the cache=] with |report| and return"?

File: index.bs [hunk 12]
```diff
@@ -2132,26 +2132,29 @@ and a [=header list=] |newHeaders| (defaults to an empty [=list=]):
...
-1. [=Queue a task=] to [=fetch=] |request| with [=fetch/processResponse=] being these steps:
-    1. If |report| is an:
-        <dl class="switch">
-        <dt>[=event-level report=]</dt>
-        <dd>[=Queue a task=] to [=list/remove=] |report| from the [=event-level report cache=].</dd>
-
-        <dt>[=aggregatable report=]</dt>
-        <dd>[=Queue a task=] to [=list/remove=] |report| from the [=aggregatable report cache=].</dd>
-        </dl>
+1. [=Queue a task=] to [=fetch=] |request| with [=fetch/processResponse=] being [=remove a report from the cache=].
```
Comment (alexmturner) [inline index.bs:2157]: ditto

**Comment by alexmturner** [comment]
LGTM % two nits :)

**Review by alexmturner — APPROVED** [review]

File: index.bs [hunk 13]
```diff
@@ -1869,6 +1867,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+1. If |report|'s [=aggregatable report/trigger debug key=] is null, return <strong>disabled</strong>.
+1. Return <strong>enabled</strong>.
+
+<h3 id="obtain-aggregatable-report-shared-info">Obtaining an aggregatable report's shared info</h3>
+
+An [=aggregatable report=] |report|'s <dfn for="aggregatable report">shared info</dfn> is the result
+of running the following steps:
+
+1. Let |reportingOrigin| be |report|'s [=aggregatable report/reporting endpoint=].
+1. [=Assert=]: |reportingOrigin| is a [=potentially trustworthy origin=].
```
Comment (apasel422) [inline index.bs:1890]: I don't think we need this assertion anymore, as the reporting endpoint is a suitable origin.

File: index.bs [hunk 14]
```diff
@@ -1869,6 +1867,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+    : "`version`"
+    :: A [=string=], API version.
+
+1. If |report|'s [=aggregatable report/debug mode=] is <strong>enabled</strong>,
+    set |sharedInfo|["`debug_mode`"] to "`enabled`".
+1. Return the [=string=] resulting from executing [=serialize an infra value to a json string=] on |sharedInfo|.
+
+<h3 id="obtain-aggregatable-report-aggregation-service-payloads">Obtaining an aggregatable report's aggregation service payloads</h3>
+
+To <dfn>obtain the public key for encryption</dfn>, return a user agent-determined public key asynchronously,
```
Comment (apasel422) [inline index.bs]:
```suggestion
To <dfn>obtain the public key for encryption</dfn>, return a user-agent-determined public key asynchronously,
```

File: index.bs [hunk 15]
```diff
@@ -1869,6 +1867,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+or an error in the event that the user agent failed to obtain the public key.
+
+Note: The user agent might enforce weekly key rotation. If there are multiple keys, the user agent
+might independently pick a key uniformly at random for every encryption operation.
+The key should be uniquely identifiable by an identifier.
```
Comment (apasel422) [inline index.bs:1920]: The "by an identifier" here seems redundant.

File: index.bs [hunk 16]
```diff
@@ -1869,6 +1867,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+1. Set up [[RFC9180|HPKE]] [[RFC9180#name-encryption-to-a-public-key|sender's context]]
+    with |pkR| and |info|.
+1. Return the [=byte sequence=] or an error resulting from [[RFC9180#name-encryption-and-decryption|encrypting]]
+    |plaintext| with the [[RFC9180#name-encryption-to-a-public-key|sender's context]].
+
+To <dfn>obtain the aggregation service payloads</dfn> given an [=aggregatable report=] |report|,
+run the following steps:
+
+1. Let |pkR| be the result of running [=obtain the public key for encryption=].
+1. If |pkR| is an error, return the error.
```
Comment (apasel422) [inline index.bs:1959]:
```suggestion
1. If |pkR| is an error, return |pkR|.
```

File: index.bs [hunk 17]
```diff
@@ -1869,6 +1867,111 @@ To <dfn>serialize [=event-level report/attribution destinations=]</dfn> |destina
...
+1. Let |encryptedPayload| be the result of running [=obtain the encrypted payload=] with |report| and |pkR|.
+1. If |encryptedPayload| is an error, return the error.
```
Comment (apasel422) [inline index.bs:1961]:
```suggestion
1. If |encryptedPayload| is an error, return |encryptedPayload|.
```

File: index.bs [hunk 18]
```diff
@@ -1905,12 +2008,14 @@ To <dfn>serialize an [=event-level report=]</dfn> |report|, run the following st
 To <dfn>serialize an [=aggregatable report=] </dfn> |report|, run the following steps:
 
 1. [=Assert=]: |report|'s [=aggregatable report/effective attribution destination=] is not the [=opaque origin=].
+1. Let |aggregationServicePayloads| be the result of running [=obtain the aggregation service payloads=].
+1. If |aggregationServicePayloads| is an error, return the error.
```
Comment (apasel422) [inline index.bs:2012]:
```suggestion
1. If |aggregationServicePayloads| is an error, return |aggregationServicePayloads|.
```

File: index.bs [hunk 19]
```diff
@@ -2029,25 +2132,29 @@ and a [=header list=] |newHeaders| (defaults to an empty [=list=]):
 
 This algorithm constructs a [=request=] and attempts to deliver it to a [=suitable origin=].
 
+To <dfn>remove a report from the cache </dfn> given an [=attribution report=] |report|:
```
Comment (apasel422) [inline index.bs:2135]:
```suggestion
To <dfn>remove a report from the cache</dfn> given an [=attribution report=] |report|:
```

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by apasel422 — APPROVED** [review]

**Pull request merged**